src/klein/test/typing_app.py

```diff
@@ -71,3 +80,36 @@ def returnsDeferredTag(self, request: IRequest) -> KleinRenderable:
     @router.route("/deferred-none")
     def returnsDeferredNone(self, request: IRequest) -> KleinRenderable:
         return succeed(None)
+
+    # Ensure that same return object types for a route are valid in async
+    # methods.
+
+    @router.route("/async-object")
+    async def asyncReturnsObject(self, request: IRequest) -> KleinRenderable:
```

Comment:
`-> KleinRenderable` isn't doing what you think here, because `KleinRenderable` includes both `KleinSynchronousRenderable` and `Awaitable[KleinSynchronousRenderable]`, and you can't return both of those here; due to the implicit modification of `async def`, the *actual* return type of this function becomes (handwaving a bit over the `Coroutine` junk) `Awaitable[Union[KleinSynchronousRenderable, Awaitable[KleinSynchronousRenderable]]`, which is gibberish.

What you actually want to do with these methods is to say, for example, "`-> str`", which will implicitly match against `-> KleinRenderable`.

Reply:
OK that works for cases where I know the specific type I'm returning.  What if I'm returning the result of something that gives me a `KleinRenderable`?

Presumably I will have awaited on it if appropriate, so I agree that maybe this should be `KleinSynchronousRenderable`, though that's currently not public.

That is, it should be possible to type this generically.

Reply:
If I use `KleinSynchronousRenderable`, I at least get passing for these, but it doesn't catch `asyncReturnsObject` as invalid… so not perfect, but better.